### PR TITLE
Save & eval from best-val checkpoint (cross-dataset)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The trainer currently saves only the **final** epoch checkpoint and reports final-epoch metrics. Validation loss is not monotone — transient minima appear mid-training and are then erased by later epochs. Evidence from PR #2895 confirms this directly: TandemFoil hit val=25.459 at epoch 109 and AirfRANS hit val=0.000371 at epoch 221, both better than the final checkpoint, both lost. Implementing best-checkpoint saving will capture these minima and lower all four dataset metrics without changing any model architecture or hyperparameter.

This is classified **CRITICAL CODE CHANGE / Priority: HIGH** in the research state and directly aligned with human directive Issue #3020 ("best-checkpoint test eval mandatory").

## Instructions

Modify `target/icml2026/train.py` as follows:

### 1. Track best validation metric per dataset

At the start of training (after initializing the model), add:
```python
best_val_metric = float('inf')
best_checkpoint_path = os.path.join(args.output_dir, "checkpoint_best.pt")
```

### 2. Save best checkpoint during validation

After computing the primary validation metric each epoch, add:
```python
if val_primary_metric < best_val_metric:
    best_val_metric = val_primary_metric
    torch.save({
        'epoch': epoch,
        'model_state_dict': model.state_dict(),
        'ema_state_dict': ema.state_dict() if ema is not None else None,
        'val_metric': val_primary_metric,
    }, best_checkpoint_path)
    wandb.log({'best_val_metric': best_val_metric, 'best_epoch': epoch})
```

The `val_primary_metric` should be whatever metric is currently used to determine the best checkpoint (primary validation loss / surface MAE). If multiple metrics are tracked, use the same one currently used for LR scheduling.

### 3. Evaluate from best checkpoint at end of training

At the very end of training (after the final epoch loop), add:
```python
if os.path.exists(best_checkpoint_path):
    print(f"Loading best checkpoint (val={best_val_metric:.6f}) for final evaluation")
    ckpt = torch.load(best_checkpoint_path)
    model.load_state_dict(ckpt['model_state_dict'])
    if ema is not None and ckpt.get('ema_state_dict') is not None:
        ema.load_state_dict(ckpt['ema_state_dict'])
        ema.copy_to(model.parameters())  # apply EMA weights
    # run final eval pass and log as 'best_checkpoint_val_metric'
    final_best_metrics = run_validation(model, ...)
    wandb.log({'final_from_best_ckpt/<metric>': v for metric, v in final_best_metrics.items()})
```

Adapt to the actual function names in the trainer. The key requirement is: **the final reported metrics come from the best-val checkpoint, not the last epoch**.

### 4. Run cross-dataset validation

Run all four datasets with the baseline config for each, so we can directly compare best-checkpoint vs. final-checkpoint metrics:

**TandemFoil** (baseline PR #2924):
```bash
cd "target/icml2026" && python train.py \
  --dataset tandemfoil \
  --lr 1.25e-4 --optimizer lion --grad-clip 0.5 --ema-decay 0.999 \
  --num-layers 3 --hidden-dim 192 --num-heads 3 \
  --wandb_group best-ckpt-cross-dataset \
  --wandb_name best-ckpt-tandemfoil
```

**TandemFoil Paper** (baseline PR #2979):
```bash
cd "target/icml2026" && python train.py \
  --dataset tandemfoil_paper \
  --lr 5e-4 --optimizer adamw --scheduler cosine --T-max 150 --ema-decay 0.999 \
  --num-layers 3 --hidden-dim 192 \
  --wandb_group best-ckpt-cross-dataset \
  --wandb_name best-ckpt-tandemfoil-paper
```

**AirfRANS** (baseline PR #2951):
```bash
cd "target/icml2026" && python train.py \
  --dataset airfrans \
  --lr 6e-4 --optimizer adamw --scheduler cosine --T-max 50 --no-use-ema \
  --num-layers 2 --hidden-dim 256 --num-heads 4 --fourier-features \
  --wandb_group best-ckpt-cross-dataset \
  --wandb_name best-ckpt-airfrans
```

**DrivAerML** (baseline PR #2898):
```bash
cd "target/icml2026" && python train.py \
  --dataset drivaerml \
  --lr 5e-4 --optimizer adamw --scheduler cosine --T-max 30 --no-use-ema \
  --num-layers 4 --hidden-dim 512 --num-heads 8 --fourier-features \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group best-ckpt-cross-dataset \
  --wandb_name best-ckpt-drivaerml
```

### 5. Report both final and best-checkpoint metrics

In your PR comment, report a table with both values side-by-side so we can measure the delta:

| Dataset | Final-ckpt val | Best-ckpt val | Best epoch | Delta |
|---------|---------------|---------------|------------|-------|
| TandemFoil | ... | ... | ... | ... |
| TandemFoil Paper | ... | ... | ... | ... |
| AirfRANS | ... | ... | ... | ... |
| DrivAerML | ... | ... | ... | ... |

## Baseline

Current best validation metrics (from BASELINE.md):

| Dataset | Best val metric | PR | Config |
|---------|----------------|----|--------|
| TandemFoil | 22.537 | #2924 | Lion lr=1.25e-4, gc=0.5, EMA=0.999, 3L/192d/3H |
| AirfRANS | 0.000482 | #2951 | AdamW lr=6e-4, T_max=50, 2L/256d/4H, no-EMA, Fourier |
| DrivAerML | 3.997% | #2898 | AdamW lr=5e-4, T_max=30, no-EMA, Fourier, 4L/512d/8H |
| TandemFoil Paper | 0.00434 | #2979 | AdamW lr=5e-4, T_max=150, no-EMA, 3L/192d |

**DrivAerML external target: <3.71%** (AB-UPT benchmark). Current gap: 1.08x.

Note from PR #2895: transient minima TF=25.459 at epoch 109 and AF=0.000371 at epoch 221 were observed but lost due to final-checkpoint-only saving. This experiment tests whether best-checkpoint saving can recover those (or better) values.

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
